### PR TITLE
fix(Core/GO): failed attempt

### PR DESCRIPTION
**Pull request by Helias**
## Changes Proposed:
-  ported from https://github.com/TrinityCore/TrinityCore/pull/21039 (related TC issue: https://github.com/TrinityCore/TrinityCore/issues/21038)
- ty @Yehonal 

## Issues Addressed:
- Closes https://github.com/chromiecraft/chromiecraft/issues/9

## Tests Performed:
- no test, only build

## How to Test the Changes:
`Just try to mine some Copper Vein, and interrupt your casts.` ( https://github.com/chromiecraft/chromiecraft/issues/9 )

## Target Branch(es):
- [x] Master

 
## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

### CODE REVIEW

**Review by Yehonal — APPROVED** [review]

**Review by FrancescoBorzi — APPROVED** [review]

**Review by Si1ker — APPROVED** [review]
Already tested in https://github.com/chromiecraft/chromiecraft/issues/9

edit: dismiss this.

**Comment by FrancescoBorzi** [comment]
wait @Si1ker , who tested this? I'm afraid that user just misunderstood and assumed the fix was already live in the server...

**Comment by Si1ker** [comment]
image
image

@FrancescoBorzi Thanks for hold my back, I thought it was implemented right away in ChromieCraft, my bad.
@Helias  Seems the normal thing happened in TrinityCore... nobody tested that and just merged it into the core.

Still failing, if there should not be a fail opportunity at all, then even I have level 1 of the profession, I must not get a failed attempt.

I did two tests:

- My custom core (1-month-old == AC): Test failed.
- 
- Another with a literally out-of-box core: Test failed.
1ed80e95 (HEAD -> main, origin/master, origin/HEAD) fix(Core/Guild): Implement gender in guild
No modules, clean build, ReadWithDeb x64, windows

Here is a proof of how my Spell.cpp is after applying the changes proposed:

```cpp
            case SPELL_EFFECT_SKINNING:
                {
                    if (m_caster->GetTypeId() != TYPEID_PLAYER || !m_targets.GetUnitTarget() || m_targets.GetUnitTarget()->GetTypeId() != TYPEID_UNIT)
                        return SPELL_FAILED_BAD_TARGETS;

                    if (!(m_targets.GetUnitTarget()->GetUInt32Value(UNIT_FIELD_FLAGS) & UNIT_FLAG_SKINNABLE))
                        return SPELL_FAILED_TARGET_UNSKINNABLE;

                    Creature* creature = m_targets.GetUnitTarget()->ToCreature();
                    if (!creature->IsCritter() && !creature->loot.isLooted())
                        return SPELL_FAILED_TARGET_NOT_LOOTED;

                    uint32 skill = creature->GetCreatureTemplate()->GetRequiredLootSkill();

                    int32 skillValue = m_caster->ToPlayer()->GetSkillValue(skill);
                    int32 TargetLevel = m_targets.GetUnitTarget()->getLevel();
                    int32 ReqValue = (skillValue < 100 ? (TargetLevel - 10) * 10 : TargetLevel * 5);
                    if (ReqValue > skillValue)
                        return SPELL_FAILED_LOW_CASTLEVEL;

                    break;
                }
```

and...

```cpp
                    // check lock compatibility
                    SpellCastResult res = CanOpenLock(i, lockId, skillId, reqSkillValue, skillValue);
                    if (res != SPELL_CAST_OK)
                        return res;

                    // chance for fail at lockpicking attempt
                    // second check prevent fail at rechecks
                    if (skillId != SKILL_NONE && (!m_selfContainer || ((*m_selfContainer) != this)))
                    {
                        bool canFailAtMax = skillId == SKILL_LOCKPICKING;

                        // chance for failure in orange lockpick
                        if ((canFailAtMax || skillValue < sWorld->GetConfigMaxSkillValue()) && reqSkillValue > irand(skillValue - 25, skillValue + 37))
                            return SPELL_FAILED_TRY_AGAIN;
                    }
                    break;
```

**Comment by FrancescoBorzi** [comment]
gj @Si1ker !

**Review by Wowhungarian — COMMENTED** [review]
This commit has not solved the problem neither on TC, nor on AC. :(

**Comment by Helias** [comment]
If I will have enough time I will investigate on it

File: src/server/game/Spells/Spell.cpp [hunk 1]
```diff
@@ -5992,13 +5986,13 @@ SpellCastResult Spell::CheckCast(bool strict)
...
+                    // chance for fail at lockpicking attempt
                     // second check prevent fail at rechecks
                     if (skillId != SKILL_NONE && (!m_selfContainer || ((*m_selfContainer) != this)))
                     {
-                        bool canFailAtMax = skillId != SKILL_HERBALISM && skillId != SKILL_MINING;
+                        bool canFailAtMax = skillId == SKILL_LOCKPICKING;
 
-                        // chance for failure in orange gather / lockpick (gathering skill can't fail at maxskill)
+                        // chance for failure in orange lockpick
                         if ((canFailAtMax || skillValue < sWorld->GetConfigMaxSkillValue()) && reqSkillValue > irand(skillValue - 25, skillValue + 37))
```
Comment (Yehonal) [inline src/server/game/Spells/Spell.cpp:5996]: The condition here must be && not ||

**Review by Yehonal — CHANGES_REQUESTED** [review]

File: src/server/game/Spells/Spell.cpp (hunk 1 above)
Comment (Yehonal) [reply src/server/game/Spells/Spell.cpp:5996]: wait, more rework needed. I'll prepare a code snippet for it

Comment (Yehonal) [reply src/server/game/Spells/Spell.cpp:5996]: Here you are, this should be the correct condition for this part

```
                      if (skillId == SKILL_LOCKPICKING && reqSkillValue > irand(skillValue - 25, skillValue + 37))
                          return SPELL_FAILED_TRY_AGAIN;
```

Comment (Yehonal) [reply src/server/game/Spells/Spell.cpp:5996]: Also remember to remove` bool canFailAtMax = skillId == SKILL_LOCKPICKING;`

**Review by Yehonal — CHANGES_REQUESTED** [review]

**Review by Wowhungarian — APPROVED** [review]
It seems to be working with @Yehonal 's solution. Thank you! :)

File: src/server/game/Spells/Spell.cpp (hunk 1 above)
Comment (Helias) [reply src/server/game/Spells/Spell.cpp]: done

**Comment by Helias** [comment]
PR updated :wink:

**Review by Wowhungarian — APPROVED** [review]

**Comment by Helias** [comment]
@Wowhungarian @Yehonal, from your point of view this PR can be set as "To Be Merged", does it solve the issue https://github.com/chromiecraft/chromiecraft/issues/9 ?

**Review by Yehonal — APPROVED** [review]

**Comment by Yehonal** [comment]
approved ;)

**Comment by Wowhungarian** [comment]
@Helias Yes it does. I also did some test with it. Failed attempt message has gone from Mining, Skinning and Herbalism. And it remained as expected at Lockpicking on rev.: db2c3ca5b47c357195e8758b4ac9163700847b9e .

Mining:

WoWScrnShot_122720_191636

Skinning:

WoWScrnShot_122720_191543

Herbalism:

WoWScrnShot_122720_191848

Lockpicking:

WoWScrnShot_122720_200021

**Comment by Yehonal** [comment]
Thanks for the tests! I guess we can merge it

**Comment by FrancescoBorzi** [comment]
@Yehonal feel free to update the labels in such cases :)

**Comment by Helias** [comment]
Thanks a lot to everyone :heart:

**Comment by Wowhungarian** [comment]
There is one more thing I need to mention. Lockpicking fails happens extremly rarely, compare to the fail rate, that mining had. Is that normal?

**Comment by Yehonal** [comment]
the rand method is the same, probably just luck :)

**Pull request merged**